<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_51) on Wed Oct 01 16:41:56 BST 2014 -->
<title>NucleobaseAndReadFactory</title>
<meta name="date" content="2014-10-01">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="NucleobaseAndReadFactory";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../artificialFastqGenerator/package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/NucleobaseAndReadFactory.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator"><span class="strong">Prev Class</span></a></li>
<li><a href="../artificialFastqGenerator/QualityScoreAndErrorGenerator.html" title="class in artificialFastqGenerator"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?artificialFastqGenerator/NucleobaseAndReadFactory.html" target="_top">Frames</a></li>
<li><a href="NucleobaseAndReadFactory.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">artificialFastqGenerator</div>
<h2 title="Class NucleobaseAndReadFactory" class="title">Class NucleobaseAndReadFactory</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>artificialFastqGenerator.NucleobaseAndReadFactory</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">NucleobaseAndReadFactory</span>
extends java.lang.Object</pre>
<div class="block">An instance of this class can be used to create new Nucleobase and Read objects.
 
 Copyright (C) 2012 The Institute of Cancer Research (ICR).

 This file is part of ArtificialFastqGenerator v1.0.0.
 
 ArtificialFastqGenerator is free software: you can redistribute it and/or modify it under the terms of the GNU General
 Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any
 later version.
 
 This program is distributed in the hope that it will be useful but WITHOUT ANY WARRANTY; without even the implied warranty
 of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 
 You should have received a copy of the GNU Public License along with this
 program. If not, see <http://www.gnu.org/licenses/>
 
 Authour's contact email: Matthew.Frampton@icr.ac.uk</div>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#NucleobaseAndReadFactory(int, int, int, boolean, boolean, java.lang.String, java.lang.String)">NucleobaseAndReadFactory</a></strong>(int&nbsp;startingXCoord,
                        int&nbsp;startingYCoord,
                        int&nbsp;readLength,
                        boolean&nbsp;useRealQualityScores,
                        boolean&nbsp;simulateErrorInRead,
                        java.lang.String&nbsp;fastq1ForQualityScoresPath,
                        java.lang.String&nbsp;fastq2ForQualityScoresPath)</code>
<div class="block">Initialises a new NucleobaseAndReadFactory object.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a></code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#createNucleobase(char)">createNucleobase</a></strong>(char&nbsp;genotype)</code>
<div class="block">Create a Nucleobase object.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#createPairedEndReads(java.util.ArrayList, java.util.ArrayList)">createPairedEndReads</a></strong>(java.util.ArrayList&lt;<a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a>&gt;&nbsp;leftEndSequence,
                    java.util.ArrayList&lt;<a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a>&gt;&nbsp;rightEndSequence)</code>
<div class="block">Create a pair of reads.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#getNumACGTBasesCalled()">getNumACGTBasesCalled</a></strong>()</code>
<div class="block">Get the number of ACGT nucleobases that have been called.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#getNumBaseCallErrors()">getNumBaseCallErrors</a></strong>()</code>
<div class="block">Get the number of base call errors that have been simulated.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#getNumNucleobasesCreated()">getNumNucleobasesCreated</a></strong>()</code>
<div class="block">Get the number of nucleobases that have been created.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#getNumReadsCreated()">getNumReadsCreated</a></strong>()</code>
<div class="block">Get the number of reads that have been created.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#logOverallErrorStats()">logOverallErrorStats</a></strong>()</code>
<div class="block">Log the overall error statistics.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../artificialFastqGenerator/NucleobaseAndReadFactory.html#resetFieldsForNewChromosome()">resetFieldsForNewChromosome</a></strong>()</code>
<div class="block">Reset fields ready to process a new chromosome.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="NucleobaseAndReadFactory(int, int, int, boolean, boolean, java.lang.String, java.lang.String)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>NucleobaseAndReadFactory</h4>
<pre>public&nbsp;NucleobaseAndReadFactory(int&nbsp;startingXCoord,
                        int&nbsp;startingYCoord,
                        int&nbsp;readLength,
                        boolean&nbsp;useRealQualityScores,
                        boolean&nbsp;simulateErrorInRead,
                        java.lang.String&nbsp;fastq1ForQualityScoresPath,
                        java.lang.String&nbsp;fastq2ForQualityScoresPath)</pre>
<div class="block">Initialises a new NucleobaseAndReadFactory object.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>startingXCoord</code> - - the starting X coordinate.</dd><dd><code>startingYCoord</code> - - the starting Y coordinate.</dd><dd><code>readLength</code> - - the lenght of each read.</dd><dd><code>useRealQualityScores</code> - - true if using quality scores from an existing fastq file, else false.</dd><dd><code>simulateErrorInRead</code> - - true if simulating error in reads, else false.</dd><dd><code>fastq1ForQualityScoresPath</code> - - the 1st fastq file whose quality scores are used.</dd><dd><code>fastq2ForQualityScoresPath</code> - - the 2nd fastq file whose quality scores are used.</dd></dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="createNucleobase(char)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createNucleobase</h4>
<pre>public&nbsp;<a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a>&nbsp;createNucleobase(char&nbsp;genotype)</pre>
<div class="block">Create a Nucleobase object.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>genotype</code> - - the nucleobase's genotype.</dd>
<dt><span class="strong">Returns:</span></dt><dd>newNucleobase - a new nucleobase object.</dd></dl>
</li>
</ul>
<a name="createPairedEndReads(java.util.ArrayList, java.util.ArrayList)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createPairedEndReads</h4>
<pre>public&nbsp;void&nbsp;createPairedEndReads(java.util.ArrayList&lt;<a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a>&gt;&nbsp;leftEndSequence,
                        java.util.ArrayList&lt;<a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator">Nucleobase</a>&gt;&nbsp;rightEndSequence)</pre>
<div class="block">Create a pair of reads.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>leftEndSequence</code> - - the sequence of nucleobases for the left read.</dd><dd><code>rightEndSequence</code> - - the sequence of nucleobases for the right read.</dd></dl>
</li>
</ul>
<a name="getNumNucleobasesCreated()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumNucleobasesCreated</h4>
<pre>public&nbsp;long&nbsp;getNumNucleobasesCreated()</pre>
<div class="block">Get the number of nucleobases that have been created.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>numNucleobasesCreated - the number of Nucleobase objects that have been created so far.</dd></dl>
</li>
</ul>
<a name="getNumReadsCreated()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumReadsCreated</h4>
<pre>public&nbsp;long&nbsp;getNumReadsCreated()</pre>
<div class="block">Get the number of reads that have been created.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>numReadsCreated - the number of Read objects that have been created so far.</dd></dl>
</li>
</ul>
<a name="getNumBaseCallErrors()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumBaseCallErrors</h4>
<pre>public&nbsp;long&nbsp;getNumBaseCallErrors()</pre>
<div class="block">Get the number of base call errors that have been simulated.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>numBaseCallErrors - the number of base call errors that have been simulated.</dd></dl>
</li>
</ul>
<a name="getNumACGTBasesCalled()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumACGTBasesCalled</h4>
<pre>public&nbsp;long&nbsp;getNumACGTBasesCalled()</pre>
<div class="block">Get the number of ACGT nucleobases that have been called.</div>
<dl><dt><span class="strong">Returns:</span></dt><dd>numACGTBasesCalled - the number of ACGT nucleobases that have been called.</dd></dl>
</li>
</ul>
<a name="logOverallErrorStats()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>logOverallErrorStats</h4>
<pre>public&nbsp;void&nbsp;logOverallErrorStats()</pre>
<div class="block">Log the overall error statistics.</div>
</li>
</ul>
<a name="resetFieldsForNewChromosome()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>resetFieldsForNewChromosome</h4>
<pre>public&nbsp;void&nbsp;resetFieldsForNewChromosome()</pre>
<div class="block">Reset fields ready to process a new chromosome.</div>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../artificialFastqGenerator/package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/NucleobaseAndReadFactory.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../artificialFastqGenerator/Nucleobase.html" title="class in artificialFastqGenerator"><span class="strong">Prev Class</span></a></li>
<li><a href="../artificialFastqGenerator/QualityScoreAndErrorGenerator.html" title="class in artificialFastqGenerator"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?artificialFastqGenerator/NucleobaseAndReadFactory.html" target="_top">Frames</a></li>
<li><a href="NucleobaseAndReadFactory.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
